### ISSUE TITLE
Fallback DNS for local IP(s)

### ISSUE THREAD

**Issue by aviv926** [celzero/rethink-app#1745]
I use Nextdns in order to rewrite my server address when I'm at home to a private address, that is, mysite.example.com has become 192.168.1.2 with a port forwarded (through the router) to the requested service which also appears public at the same address (mysite.example.com)

The thing is, when I'm out of the house, the address 192.168.1.2 is not available of course, so I would expect that when for several attempts to locate 192.168.1.2 by the main DNS server without success, the Fallback DNS will come into effect in order to retrieve the true address of the server.

what do you think Is it applicable?
Will there be in-app support for cases of self-hosting and moving between different networks?
For example, make Rethink behave in an X way when connected to a WIFI network / a specific WIFI network and in a Y way when operating on cellular networks.

**Comment by ignoramous** [comment]
> I use Nextdns in order to rewrite my server address when I'm at home to a private address

NextDNS setup at the router? If so, *System DNS* should use NextDNS when you're connected to that router. Is that not what you see happen?

> Will there be in-app support for cases of self-hosting and moving between different networks?

It wouldn't take us long to implement this, but the UI / UX is already a nightmare and adding more knobs is only adding more clutter.

Possible duplicate?
- https://github.com/celzero/rethink-app/issues/1130

See also:
- https://github.com/celzero/rethink-app/issues/1143
- https://github.com/celzero/rethink-app/issues/1294

**Comment by aviv926** [comment]
> NextDNS setup at the router? If so, _System DNS_ should use NextDNS when you're connected to that router. Is that not what you see happen?

No, I use NextDns in DoH with a customized profile in Rethink, so automatically the DNS of the router is not relevant in this case, so I would expect Fallback DNS to come into action and find the IP address of the server when I am not in my home network.

> It wouldn't take us long to implement this, but the UI / UX is already a nightmare and adding more knobs is only adding more clutter.

That's good to hear, I hope it arrives, when it arrives... Thanks for the hard work you put into this project!
> Possible duplicate?
> 
> * [Support per-SSID DNS types #1130](https://github.com/celzero/rethink-app/issues/1130)
> 
> See also:
> 
> * [Auto-enable proxies based on network type Mobile, WiFi #1143](https://github.com/celzero/rethink-app/issues/1143)

[wgtunnel ](https://github.com/celzero/rethink-app/issues/1143#issuecomment-1978962695) sounds like a **great** option the problem is that it also needs to create a VPN connection but then Rethink won't be used (which is not something I want) I hope one day there will be a union between them or a merger of features from wgtunnel to Rethink.

> * [Privacy friendly way to turn on Wireguard depend on home/away scenario #1294](https://github.com/celzero/rethink-app/issues/1294)

**Comment by aviv926** [comment]
@ignoramous Do you think it's a problem that Fallback DNS doesn't try to resolve the domain in the scenario I wrote?

**Comment by ignoramous** [comment]
> Do you think it's a problem that Fallback DNS doesn't ...

1. *Fallback DNS* is *not* used when the actual user-preferred DNS is *selected*, even if unresponsive.
2. *Fallback DNS* is also a "bootstrap" helping resolve IPs of DoH/DoT/ODoH resolvers.
3. *Fallback DNS* is also used by Rethink for its own DNS (especially when in *Loopback* mode).

**Comment by aviv926** [comment]
> > Do you think it's a problem that Fallback DNS doesn't ...
> 
> 1. *Fallback DNS* is *not* used when the actual user-preferred DNS is *selected*, even if unresponsive.
> 2. *Fallback DNS* is also a "bootstrap" helping resolve IPs of DoH/DoT/ODoH resolvers.
> 3. *Fallback DNS* is also used by Rethink for its own DNS (especially when in *Loopback* mode).


Oh I see, actually as it is called I was sure it is used when the DNS selected by the user does not return an "answer" or when it is an unavailable address - a private address (if it is possible to make this happen understand when it was able to decode a private address and when the address is private  Not available (due to mobile data) Do you think this will be supported in future versions?

**Comment by ignoramous** [comment]
Switching DNSes based on mobile / wifi networks is tracked at #1130

> Do you think this will be supported in future versions?

Not this specific usecase but #1130 will be.

**Comment by ignoramous** [comment]
> The thing is, when I'm out of the house, the address 192.168.1.2 is not available of course, so I would expect that when for several attempts to locate 192.168.1.2 by the main DNS server without success, the Fallback DNS will come into effect in order to retrieve the true address of the server.

@aviv926 Rethink should respect upstream TTLs for domains of DNS upstreams setup in *Configure -> DNS -> Other DNS* (regardless of *Configure -> DNS -> DNS Booster* setting which enables an aggressive caching mechanism). The scenario where the domain <> IP mapping is no longer valid (that is, domain is unreachable over previously resolved IPs) should trigger Rethink to re-resolve the domain and fetch a new set of IPs again from *Configure -> Network -> Choose Boostrap DNS* (in versions prior to [`v056`](https://github.com/celzero/rethink-app/releases/v0.5.6), this was called *Choose fallback DNS*). Can you confirm if that's not the case? If so, we *want* to support this usecase (on priority) in one of the upcoming versions.

**Comment by aviv926** [comment]
> _Configure -> Network -> Choose Boostrap DNS_

Hello, I have done some further investigation now with the latest version (installation and download via the GitHub release page)

**Current configuration:**
* VPN proxy is running in advanced mode together with DNS on Split DNS mode
* Prevent DNS Leak is on
* DNS for bypass in automatic mode
* DNS configured in Rethink is nextdns with Rewrites that points a record to a private address (for example google.com points to 192.168.1.5)
* Configure -> Network -> Choose Boostrap DNS is set to Rethink's DNS


The application on which the experiment is being performed is Firefox Focus which is set to bypass app from all proxies mode

**Test results:**

### When I am connected to my home WIFI:
When I am connected to my home wifi  (which is configured on its router Firewall - Port Forwards for google.com points to 192.168.1.5) then it works as expected and google.com does point to 192.168.1.5

### When I get off the WIFI network:

When I get off the wifi network  and then the Firewall - Port Forwards is no longer there The address refuses to load and no valid DNS address is received from Boostrap DNS (i.e. the server is unavailable timeout)
The log does not show any relevant information other than the attempt to load 192.168.1.5 as google.com

I have also tried putting the phone into airplane mode and disabling and re-enabling the VPN but that did not solve the problem.
